### PR TITLE
CORDA-2456: Documentation around explicit upgrades

### PR DESCRIPTION
**Pull request by tudor-malene**
This PR just touches only the explicit upgrade part.

This document needs some overall rethinking.

### CODE REVIEW

File: docs/source/upgrading-cordapps.rst [hunk 1]
```diff
@@ -14,6 +14,8 @@ Release new CorDapp versions
 
 CorDapp versioning
 ------------------
+.. UPDATE - This is no longer accurate! Needs to talk about the different types of artifacts ( kernel, workflows) each versioned independently
```
Comment (mikehearn) [inline docs/source/upgrading-cordapps.rst]: can we file a ticket for this?

File: docs/source/upgrading-cordapps.rst [hunk 2]
```diff
@@ -279,25 +281,25 @@ Performing explicit contract and state upgrades
...
 1. Preserve the existing state and contract definitions
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Currently, all nodes must **permanently** keep **all** old state and contract definitions on their node's classpath
+Currently, all nodes must **permanently** keep **all** old state and contract definitions on their node's classpath if the explicit upgrade
+process was used on them.
 
-.. note:: Once the contract-code-as-an-attachment feature has been implemented, nodes will only be required to keep the
-   old state and contract definitions on their node's classpath for the duration of the upgrade
+.. note:: This requirement will go away in a future version of Corda. In Corda 4, the contract-code-as-attachment feature was implemented
+          only for "normal" transactions. `Contract Upgrade` and `Notary Change` transactions will still be executed withing the node classpath.
```
Comment (mikehearn) [inline docs/source/upgrading-cordapps.rst]: typo: within. single backticks don't do what's expected here, I think?

File: docs/source/upgrading-cordapps.rst [hunk 3]
```diff
@@ -307,11 +309,20 @@ upgraded contracts must implement the ``UpgradedContract`` interface. This inter
...
 
-The ``upgrade`` method describes how the old state type is upgraded to the new state type. When the state isn't being
-upgraded, the same state type can be used for both the old and new state type parameters.
+The ``upgrade`` method describes how the old state type is upgraded to the new state type.
 
 By default this new contract will only be able to upgrade legacy states which are constrained by the zone whitelist (see :doc:`api-contract-constraints`).
-If hash or other constraint types are used, the new contract should implement ``UpgradedContractWithLegacyConstraint``
+
+
+.. note:: The requirement for a `legacyContractConstraint` arises from the fact that when a transaction chain is verified and a `Contract Upgrade` is
```
Comment (mikehearn) [inline docs/source/upgrading-cordapps.rst]: double backticks, or no?

File: docs/source/upgrading-cordapps.rst [hunk 4]
```diff
@@ -307,11 +309,20 @@ upgraded contracts must implement the ``UpgradedContract`` interface. This inter
...
+          encountered on the back chain, the verifier wants to know that a legitimate state was transformed into the new contract. The `legacyContractConstraint` is
+          the mechanism by which this is enforced. Using it, the new contract is able to narrow down what constraint the states it is upgrading should have.
+          If a malicious party would create a fake `com.megacorp.MegaToken` state, he would not be able to use the usual `MegaToken` code as his
```
Comment (mikehearn) [inline docs/source/upgrading-cordapps.rst]: double backticks

File: docs/source/upgrading-cordapps.rst (hunk 1 above)
Comment (tudor-malene) [reply docs/source/upgrading-cordapps.rst]: Created: https://r3-cev.atlassian.net/browse/ENT-3483

**Comment by tudor-malene** [comment]
Back to you @mikehearn

**Pull request merged**